tests/apollo/test_skvbc_chaotic_startup.py

```diff
@@ -237,24 +218,114 @@ async def test_f_staggered_replicas_requesting_vc(self, bft_network, skvbc, nurs
+        await self._wait_for_less_than_f_plus_one_replicas_to_initiate_viewchange(early_replicas, bft_network)
+        print("STATUS: Early replicas started and initiated View Change.")
+
+
+        late_replicas = bft_network.all_replicas(without=early_replicas)
+        print(f"STATUS: Starting the remaining {n - f + 1} replicas.")
+        bft_network.start_replicas(late_replicas)
+
+        view = await bft_network.wait_for_view(
+            replica_id=initial_primary,
+            expected=lambda v: v == initial_primary,
+            err_msg="Make sure we are in the initial view "
+        )
+        self.assertTrue(initial_view == view)
+
+        await self._verify_the_system_is_able_to_make_progress(bft_network, skvbc, initial_primary, late_replicas)
+
+        # Verify replicas that have initiated View Change catch up on state
+        await bft_network.wait_for_state_transfer_to_start()
+        for r in early_replicas:
+            await bft_network.wait_for_state_transfer_to_stop(initial_primary,
+                                                                  r,
+                                                                  stop_on_stable_seq_num=True)
+            print("STATUS: Early replicas that have initiated View Change catch up on state.")
+
+        # We stop the current primary and let the system to install a new view, while assuming the identity of
+        # the next primary
+        bft_network.stop_replica(initial_primary)
+
+        view = await bft_network.wait_for_view(
+            replica_id=expected_next_primary,
+            expected=lambda v: v == expected_next_view,
+            err_msg="Make sure we are in the next view "
+        )
+
+        self.assertTrue(expected_next_view == view)
+
+        await trio.sleep(5)  # TODO: remove when bft_network.wait_for_view also waits for system liveness.
+
+        await self._verify_replicas_are_in_view(view, bft_network.all_replicas(without={initial_primary}), bft_network)
+        await self._verify_the_system_is_able_to_make_progress(bft_network, skvbc, expected_next_primary, bft_network.all_replicas(without={initial_primary}))
+
+    async def _verify_replicas_are_in_view(self, view, replicas, bft_network):
+        for r in replicas:
             active_view_of_replica = await self._get_gauge(r, bft_network, 'currentActiveView')
             view_of_replica = await self._get_gauge(r, bft_network, 'view')
             self.assertTrue(active_view_of_replica == view_of_replica)  # verify Replica is not requesting View Change
             self.assertTrue(active_view_of_replica == view)  # verify Replica is in the current view
 
-        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=expected_next_primary)
+    async def _verify_the_system_is_able_to_make_progress(self, bft_network, skvbc, replica_to_read_from, initial_nodes, checkpoint_num=1, verify_checkpoint_persistency=False, assert_state_transfer_not_started=False):
```

Comment:
Just a suggestion might be call this method _wait_for_replicas_to_generate_stable_checkpoint?

Reply:
done